tests: arch: improving the coverage of arch code

Zhaoningx opened the pull request:
Adding two testcases to:
Verify arch interface by calling timing APIs.
    
Verify the API k_cpu_atomic_idle() can keep the bit value
of IF in EFLAGS register before and after the call.

Signed-off-by: NingX Zhao <ningx.zhao@intel.com>

enjiamai commented:
Hi @Zhaoningx , I checked this test case pass in ACRN ehl_crb, could you please also check if this passes on up_squared and ehl_crb boards, thanks!

LeiW000 commented on tests/arch/x86/nmi/src/main.c:
```diff
@@ -65,9 +67,44 @@ void test_nmi_handler(void)
+	cycle_to_ns = arch_timing_cycles_to_ns_avg(1000, 10);
+	arch_timing_stop();
+
+	zassert_true(freq != 0, "failed to get arch mHZ");
```
It is better to change the error message like below.
"Failed to call arch_timing_freq_get_mhz()."

Zhaoningx replied:
Hi @LeiW000 , updated, thanks.

Zhaoningx commented:
> Hi @Zhaoningx , I checked this test case pass in ACRN ehl_crb, could you please also check if this passes on up_squared and ehl_crb boards, thanks!

Hi @enjiamai , I tested it on up_squared board, it works. But the ehl_crb board is hold by others, so I'm still not testing for it. Thanks.

enjiamai commented on tests/arch/x86/nmi/src/main.c:
```diff
@@ -65,9 +81,91 @@ void test_nmi_handler(void)
+	arch_timing_stop();
+
+	zassert_true(freq != 0, "Failed to call arch_timing_freq_get_mhz().");
+	zassert_true(cycle_to_ns != 0, "failed to call arch_timing_cycles_to_ns_avg");
```
Is there any other condition we could check more than 0?

Zhaoningx replied:
Hi @enjiamai, I don't know the better idea to check if the timing works, so maybe I need some time to dig into it.

dcpleung reviewed (changes requested):
The new additions are not NMI related. Please do not put those here.

Zhaoningx commented on tests/arch/x86/nmi/src/main.c:
```diff
@@ -65,9 +81,91 @@ void test_nmi_handler(void)
+#endif
+
+/**
+ * @brief Add a function to verify arch interface
```
Updated, thanks.

Zhaoningx commented:
> The new additions are not NMI related. Please do not put those here.

Hi @dcpleung , thanks for your comment, I have moved the changes to another directioy.

enjiamai commented:
> The new additions are not NMI related. Please do not put those here.

Hi @dcpleung ,  how do you think if we changes the testsuite and directory name instead of nmi? For example, arch_funcs or something like that, then we can put other arch relevant cases here besides nmi? This means we can put more testcases into this testsuite, thanks!

dcpleung commented on tests/arch/x86/arch_interface/src/main.c:
```diff
@@ -0,0 +1,101 @@
+	uint32_t freq = 0;
+	uint64_t cycle_to_ns = 0;
+
+	arch_timing_init();
+	arch_timing_start();
```
Why the need to test the arch interface for timing functions, and not the timing functions themselves? The timing functions will call into the arch timing functions, so there is no need for separate tests for these.

Zhaoningx replied:
Hi @dcpleung, I re-designed the test case for timing, could you help to review this? Thanks.

dcpleung replied:
Usage of timing functions should have been covered by `tests/benchmarks/latency_measure`. Please consolidate this into that test.

Zhaoningx replied:
OK, thanks.

dcpleung commented on tests/arch/x86/arch_interface/src/main.c:
```diff
@@ -0,0 +1,101 @@
+ *
+ * @see k_cpu_atomic_idle()
+ */
+void test_kernel_cpu_idle_no_impact_if(void)
```
What functionalities are you trying to test here? Why do you need to test if eflags changes?????

Zhaoningx replied:
I want to test that the function k_cpu_atomic_idle can record the state of the IF bit through the key, and restore the level of the IF bit to the state before the call k_cpu_atomic_idle whatever the IF bit is 1 or 0. Do you think this test case is making sense? Thanks.

dcpleung replied:
No. Because if this function does not work correctly, the atomic idle test in `tests/kernel/context` would fail or even hang.

Zhaoningx replied:
I know the test cases in `tests/kernel/context`, but it still does not cover the condition that IF bit is 0 before calling the API k_cpu_atomic_idle, it means the command `cli` is still not tested, I add this test case is to test this branch.

dcpleung replied:
So this is done only for the sake of code coverage and nothing to do with it's functionality?

Zhaoningx replied:
Yes, this is module testing, and the test case is designed by the source code logic, so no functionality here. And I will also consolidate this into `tests/kernel/context`?

Commit:
tests: arch: improving the coverage of arch code

Adding two testcases to:
Verify arch interface by calling timing APIs.

Verify the API k_cpu_atomic_idle() can keep the bit value
of IF in EFLAGS register before and after the call.

Signed-off-by: NingX Zhao <ningx.zhao@intel.com>

Zhaoningx commented:
Hi @dcpleung, you suggested me to remove the testcase of testing timing APIs, because timing APIs are tested in benchmark dirctory, and I double check it, it works well, but I have an open for this, the testcase of latency_measure can not be used to generate code coverage. the error log as below:
```
timing init subsys
arch_timing_x86_init
START - Time Measurement
Timing results: Clock frequency: 999 MHz
Average thread context switch using yield                   :   54428 cycles ,    54428 ns
 Error: tick occurred
Switch from ISR back to interrupted thread                  :   26944 cycles ,    26944 ns
Time from ISR to executing a different thread               :   23616 cycles ,    23616 ns
Time to create a thread (without start)                     :   16512 cycles ,    16512 ns
Time to start a thread                                      :   62720 cycles ,    62720 ns
Time to suspend a thread                                    :   60224 cycles ,    60224 ns
Time to resume a thread                                     :   61632 cycles ,    61632 ns
Time to abort a thread (not running)                        :   10272 cycles ,    10272 ns
Average semaphore signal time                               :   21282 cycles ,    21282 ns
Average semaphore test time                                 :    7874 cycles ,     7874 ns
Semaphore take time (context switch)                        :   63040 cycles ,    63040 ns
Semaphore give time (context switch)                        :   98368 cycles ,    98368 ns
Average time to lock a mutex                                :    8802 cycles ,     8802 ns
Average time to unlock a mutex                              :   29009 cycles ,    29010 ns
Average time for heap malloc                                :  116128 cycles ,   116128 ns
Average time for heap free                                  :  117728 cycles ,   117728 ns
===================================================================
PROJECT EXECUTION FAILED

```
the command is `twister -p qemu_x86 -T tests/benchmarks/latency_measure/ -C`

is it normal? And need I to report a bug? @dcpleung  do you have any comments here? Thanks.

dcpleung commented:
`latency_measure` is not failing on my side with timing functions turned on. Here are what I have:

`qemu_x86`:

```
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..*** Booting Zephyr OS build v2.7.99-2330-g903c78c207fc  ***
START - Time Measurement
Timing results: Clock frequency: 999 MHz
Average thread context switch using yield                   :   11558 cycles ,    11558 ns
Average context switch time between threads (coop)          :   21053 cycles ,    21053 ns
Switch from ISR back to interrupted thread                  :    4928 cycles ,     4928 ns
Time from ISR to executing a different thread               :    3008 cycles ,     3008 ns
Time to create a thread (without start)                     :    4224 cycles ,     4224 ns
Time to start a thread                                      :   10784 cycles ,    10784 ns
Time to suspend a thread                                    :   10400 cycles ,    10400 ns
Time to resume a thread                                     :   10688 cycles ,    10688 ns
Time to abort a thread (not running)                        :    1536 cycles ,     1536 ns
Average semaphore signal time                               :    3424 cycles ,     3424 ns
Average semaphore test time                                 :    1344 cycles ,     1344 ns
Semaphore take time (context switch)                        :   12736 cycles ,    12736 ns
Semaphore give time (context switch)                        :   17568 cycles ,    17568 ns
Average time to lock a mutex                                :    1632 cycles ,     1632 ns
Average time to unlock a mutex                              :    4738 cycles ,     4738 ns
Average time for heap malloc                                :   19136 cycles ,    19136 ns
Average time for heap free                                  :   15968 cycles ,    15968 ns
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

`qemu_x86_64`:

```
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..*** Booting Zephyr OS build v2.7.99-2330-g903c78c207fc  ***
START - Time Measurement
Timing results: Clock frequency: 999 MHz
Average thread context switch using yield                   :    9060 cycles ,     9060 ns
Average context switch time between threads (coop)          :    9503 cycles ,     9503 ns
Switch from ISR back to interrupted thread                  :   14240 cycles ,    14240 ns
Time from ISR to executing a different thread               :    9664 cycles ,     9664 ns
Time to create a thread (without start)                     :    3968 cycles ,     3968 ns
Time to start a thread                                      :   12064 cycles ,    12064 ns
Time to suspend a thread                                    :   12640 cycles ,    12640 ns
Time to resume a thread                                     :   12096 cycles ,    12096 ns
Time to abort a thread (not running)                        :    2208 cycles ,     2208 ns
Average semaphore signal time                               :    8960 cycles ,     8960 ns
Average semaphore test time                                 :    2048 cycles ,     2048 ns
Semaphore take time (context switch)                        :   13472 cycles ,    13472 ns
Semaphore give time (context switch)                        :   18400 cycles ,    18400 ns
Average time to lock a mutex                                :    3072 cycles ,     3072 ns
Average time to unlock a mutex                              :    9283 cycles ,     9283 ns
Average time for heap malloc                                :   13056 cycles ,    13056 ns
Average time for heap free                                  :    7872 cycles ,     7872 ns
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Zhaoningx commented:
> `latency_measure` is not failing on my side with timing functions turned on. Here are what I have:
> 
> `qemu_x86`:
> 
> ```
> SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
> Booting from ROM..*** Booting Zephyr OS build v2.7.99-2330-g903c78c207fc  ***
> START - Time Measurement
> Timing results: Clock frequency: 999 MHz
> Average thread context switch using yield                   :   11558 cycles ,    11558 ns
> Average context switch time between threads (coop)          :   21053 cycles ,    21053 ns
> Switch from ISR back to interrupted thread                  :    4928 cycles ,     4928 ns
> Time from ISR to executing a different thread               :    3008 cycles ,     3008 ns
> Time to create a thread (without start)                     :    4224 cycles ,     4224 ns
> Time to start a thread                                      :   10784 cycles ,    10784 ns
> Time to suspend a thread                                    :   10400 cycles ,    10400 ns
> Time to resume a thread                                     :   10688 cycles ,    10688 ns
> Time to abort a thread (not running)                        :    1536 cycles ,     1536 ns
> Average semaphore signal time                               :    3424 cycles ,     3424 ns
> Average semaphore test time                                 :    1344 cycles ,     1344 ns
> Semaphore take time (context switch)                        :   12736 cycles ,    12736 ns
> Semaphore give time (context switch)                        :   17568 cycles ,    17568 ns
> Average time to lock a mutex                                :    1632 cycles ,     1632 ns
> Average time to unlock a mutex                              :    4738 cycles ,     4738 ns
> Average time for heap malloc                                :   19136 cycles ,    19136 ns
> Average time for heap free                                  :   15968 cycles ,    15968 ns
> ===================================================================
> PROJECT EXECUTION SUCCESSFUL
> ```
> 
> `qemu_x86_64`:
> 
> ```
> SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
> Booting from ROM..*** Booting Zephyr OS build v2.7.99-2330-g903c78c207fc  ***
> START - Time Measurement
> Timing results: Clock frequency: 999 MHz
> Average thread context switch using yield                   :    9060 cycles ,     9060 ns
> Average context switch time between threads (coop)          :    9503 cycles ,     9503 ns
> Switch from ISR back to interrupted thread                  :   14240 cycles ,    14240 ns
> Time from ISR to executing a different thread               :    9664 cycles ,     9664 ns
> Time to create a thread (without start)                     :    3968 cycles ,     3968 ns
> Time to start a thread                                      :   12064 cycles ,    12064 ns
> Time to suspend a thread                                    :   12640 cycles ,    12640 ns
> Time to resume a thread                                     :   12096 cycles ,    12096 ns
> Time to abort a thread (not running)                        :    2208 cycles ,     2208 ns
> Average semaphore signal time                               :    8960 cycles ,     8960 ns
> Average semaphore test time                                 :    2048 cycles ,     2048 ns
> Semaphore take time (context switch)                        :   13472 cycles ,    13472 ns
> Semaphore give time (context switch)                        :   18400 cycles ,    18400 ns
> Average time to lock a mutex                                :    3072 cycles ,     3072 ns
> Average time to unlock a mutex                              :    9283 cycles ,     9283 ns
> Average time for heap malloc                                :   13056 cycles ,    13056 ns
> Average time for heap free                                  :    7872 cycles ,     7872 ns
> ===================================================================
> PROJECT EXECUTION SUCCESSFUL
> ```

Hi @dcpleung, I think you didn't add the option `--coverage`, because I find that the case will be failed, only when I try to add `--coverage` to generate a coverage report. And @enjiamai has helped to fix this, see https://github.com/zephyrproject-rtos/zephyr/pull/41139 , so closed this PR. Thanks.
